framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,408 @@
+                }
+                // Annotation name, which is the same as the file name but with
+                // extension removed
+                String annotationName = fileName;
+                if (fileName.lastIndexOf(DOT) != -1) {
+                    annotationName = fileName.substring(0, fileName.lastIndexOf(DOT));
+                }
+
+                // Fully qualified annotation name
+                String fullyQualifiedAnnoName = qualPackageName + annotationName;
+
+                // Filter for a java file
+                if (fileName.endsWith(JAVA_SUFFIX)) {
+                    // Load the class
+                    Class<? extends Annotation> annotationClass = loadAnnotationClass(fullyQualifiedAnnoName);
+                    if (annotationClass != null) {
+                        // If successfully loaded, add to annotations set
+                        annotations.add(annotationClass);
+                    }
+                }
+            } else if (file.isDirectory()) {
+                // Descend into the directory and recursively load annotations
+                annotations.addAll(loadExternalAnnotationClassesFromDirectory(rootDirectory, file));
+            }
+        }
+
+        return annotations;
+    }
+
+    /**
+     * Loads the class indicated by the fullyQualifiedClassName, and checks to
+     * see if it is an annotation that is supported by a checker.
+     *
+     * @param fullyQualifiedClassName
+     *            the fully qualified name of the class
+     * @return the loaded annotation class if it is defined with
+     *         ElementType.TYPE_USE and is a supported annotation, null
+     *         otherwise
+     */
+    private /*@Nullable*/ Class<? extends Annotation> loadAnnotationClass(String fullyQualifiedClassName) {
+        try {
+            // load the class
+            Class<?> cls = Class.forName(fullyQualifiedClassName);
+            // ensure that the freshly loaded class is an annotation, and has
+            // the @Target annotation
+            if (cls.isAnnotation() && cls.getAnnotation(Target.class) != null) {
+                // retrieve the set of ElementTypes in the @Target
+                // meta-annotation and check to see if this annotation is
+                // supported for automatic loading
+                if (AnnotatedTypes.hasTypeTargets(cls.getAnnotation(Target.class).value())) {
+                    // if it is supported, then subclass it as an Annotation
+                    // class
+                    Class<? extends Annotation> annoClass = cls.asSubclass(Annotation.class);
+
+                    // see if the annotation is supported by a checker
+                    if (isSupportedAnnotationClass(annoClass)) {
+                        return annoClass;
+                    }
+                }
+            }
+        } catch (ClassNotFoundException e) {
```

Comment:
Try to use the minimal scopes for try/catch blocks. Here, is this only for the Class.forName in the first statement?